business/checkers/destinationrules/no_dest_checker.go

```diff
@@ -21,33 +21,63 @@ func (n NoDestinationChecker) Check() ([]*models.IstioCheck, bool) {
 	if host, ok := n.DestinationRule.GetSpec()["host"]; ok {
 		if dHost, ok := host.(string); ok {
 			fqdn := FormatHostnameForPrefixSearch(dHost, n.DestinationRule.GetObjectMeta().Namespace, n.DestinationRule.GetObjectMeta().ClusterName)
-			if versions, found := n.Services[fqdn.Service]; found {
-				if hasSubsets(n.DestinationRule) {
-					indexes := kubernetes.ValidateDestinationRulesSubsets([]kubernetes.IstioObject{n.DestinationRule}, fqdn.Service, versions)
-					for _, i := range indexes {
-						validation := models.BuildCheck("This subset is not found from the host", "error", "spec/subsets["+strconv.Itoa(i)+"]/version")
-						validations = append(validations, &validation)
-						valid = false
+			if subsets, ok := n.DestinationRule.GetSpec()["subsets"]; ok {
+				if dSubsets, ok := subsets.([]interface{}); ok {
+					// Check that each subset has a matching workload somewhere..
+					for i, subset := range dSubsets {
+						if innerSubset, ok := subset.(map[string]interface{}); ok {
+							if labels, ok := innerSubset["labels"]; ok {
+								if dLabels, ok := labels.(map[string]interface{}); ok {
+									stringLabels := make(map[string]string, len(dLabels))
+									for k, v := range dLabels {
+										if s, ok := v.(string); ok {
+											stringLabels[k] = s
+										}
+									}
+									if !n.hasMatchingWorkload(fqdn.Service, stringLabels) {
+										validation := models.BuildCheck("This subset's labels are not found from any matching host", "error", "spec/subsets["+strconv.Itoa(i)+"]")
+										validations = append(validations, &validation)
+										valid = false
+									}
+								}
+							}
+						}
 					}
+
 				}
 			} else {
-				validation := models.BuildCheck("Host doesn't have a valid service", "error", "spec/host")
-				validations = append(validations, &validation)
-				valid = false
+				if !n.hasMatchingService(fqdn.Service) {
+					validation := models.BuildCheck("This host has no matching workloads", "error", "spec/host")
+					validations = append(validations, &validation)
+					valid = false
+				}
 			}
 		}
 	}
 
 	return validations, valid
 }
 
-func hasSubsets(dr kubernetes.IstioObject) bool {
-	if subsets, ok := dr.GetSpec()["subsets"]; ok {
-		if subsetsSI, ok := subsets.([]interface{}); ok {
-			if len(subsetsSI) > 0 {
+func (n NoDestinationChecker) hasMatchingWorkload(service string, labels map[string]string) bool {
+	for _, wl := range n.WorkloadList.Workloads {
+		if service == wl.Labels["app"] {
+			for k, v := range labels {
+				wlv, found := wl.Labels[k]
+				if !found || wlv != v {
+					break
+				}
```

Comment:
This logic is missing one label, that's why it is weird looking (and wrong). Also, not sure of github, but the return true might be in the wrong indentation (I'll have to check from the editor)

Reply:
I understand that you have the whole list of workloads and with the first if (app label match) you are making sure that workload is target of that DR. otherwise, that workload doesn't apply to that DR. Once this filter is passed, then you compare all the other labels to have the full match. Is that right?
Why do you say that is missing one label?

Reply:
I see new changes since my review. Currently, this block looks OK for me.